Exclude grub packages from apt-mark hold

kreeuwijk opened the pull request:
<!--
Thank you so much for taking the time to contribute to `image-builder` ❤️

Before submitting a new PR please ensure the following:
- You have checked the open pull requests to see if there is already similar work in progress
- You have checked for current open issues matching this change to reference below

Please be patient with waiting for a review. We're a small team of contributors but try our best to get to PRs in a timely manner.

If you'd like to discuss your change with us please reach out any of the communication methods listed on the readme (https://github.com/kubernetes-sigs/image-builder#community-discussion-contribution-and-support).

-->

## Change description
<!-- What this PR does / why we need it. -->

Fixes the problem of MaaS not being able to update the `grub` packages during deployment, as documented [here](https://discourse.maas.io/t/curtin-command-fails-due-to-grub-efi-amd64-pkg-conflicts/7475/3)

<!--
If your PR include introducing new Providers or Operating systems to support please fill out the following questions.
If not, please feel free to leave blank or remove.
-->
- Is this change including a new Provider or a new OS? (y/n) __n__

mboersma reviewed (commented):
This is a specific exception for a use case that hasn't been reported as an image-builder issue. It's also not immediately clear to me how the old [MaaS bug report](https://discourse.maas.io/t/curtin-command-fails-due-to-grub-efi-amd64-pkg-conflicts/7475/3) involves image-builder: they just discuss Packer.

Can MaaS users just run `apt-mark unhold grub` before running `curtin` as a workaround?

kreeuwijk commented:
I ran into the issue myself when Canonical released an updated version of grub for Ubuntu 22.04. It was working fine before that point but afterwards MaaS deployments would fail as MaaS needs to align grub with the system architecture (legacy BIOS or UEFI), which requires adjusting the grub packages. When the grub packages are held, the process fails.

Curtin isn't helpful here, as you can only control `early_commands` and `late_commands` through curtin_userdata in MaaS. The `early_commands` run even before the image as applied to the disk and `late_commands` runs after grub has already failed.

Hence it would be better if the fix was in the image-builder. What we could do though, is split the original ansible command into 2 variants, and only run the variant that excludes grub packages when it runs for a MaaS build. There would need to be a way to identify that it's running for a MaaS build to do so. Can we set a variable in `images/capi/packer/maas/packer.json` that could be queried from the ansible task?

kreeuwijk commented:
I found a cleaner way to do it that doesn't require adding a custom role.
```
    qemu: TASK [sysprep : Pin all installed packages with apt-mark] **********************
    qemu: changed: [default]
    qemu:
    qemu: TASK [sysprep : Unpin grub packages with apt-mark for MaaS] ********************
    qemu: changed: [default]
```

AverageMarcus commented:
Oh nice! I didn't know we had a provider var 🤣 Do you know where that gets set?

/ok-to-test
/lgtm

AverageMarcus commented:
Answering my own question - the provider is specifically set in the makefile for the MaaS targets. That's nice. Maybe we should have that as a standard thing 🤔

AverageMarcus reviewed (commented):
/lgtm
/assign @mboersma @AndiDog 
/cc @glitchcrab

mboersma reviewed (approved):
/lgtm
/approve

Thanks @kreeuwijk!
